File: course1_algorithmic_toolbox/week5_dynamic_programming1/1_money_change_again/ChangeDP.java
import java.util.Scanner;

public class ChangeDP {
	private static int getChange(int money) {
		int[] minNumCoins = new int[money + 1];
		int[] coins = {1, 3, 4};

		for (int m = 1; m <= money; m++) {
			minNumCoins[m] = Integer.MAX_VALUE;
			for (int coin : coins) {
				if (m >= coin) {
					int numCoins = minNumCoins[m - coin] + 1;
					if (numCoins < minNumCoins[m]) {
						minNumCoins[m] = numCoins;
					}
				}
			}
		}
		return minNumCoins[money];
	}

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		int m = scanner.nextInt();
		System.out.println(getChange(m));

	}
}

File: course1_algorithmic_toolbox/week5_dynamic_programming1/1_money_change_again/change_dp.js
const readline = require('readline');
const rl = readline.createInterface({
	input: process.stdin,
	terminal: false
});

process.stdin.setEncoding('utf8');

const changeDP = (money) => {
	const minNumCoins = Array(money + 1).fill(0);

	const coins = [1, 3, 4];

	for (let m = 1; m <= money; m++) {
		minNumCoins[m] = Number.MAX_VALUE;
		for (const coin of coins) {
			if (m >= coin) {
				const numCoins = minNumCoins[m - coin] + 1;
				if (numCoins < minNumCoins[m]) {
					minNumCoins[m] = numCoins;
				}
			}
		}
	}
	return minNumCoins[money];

};

const readLine = (line) => {
	console.log(changeDP(parseInt(line, 10)));
	process.exit();
};

rl.on('line', readLine);

module.exports = changeDP;
